### ISSUE TITLE
Transparency and opacity should not be linked.

### ISSUE THREAD

**Issue by sarsko** [creusot-rs/creusot#987]
Currently the following:

```rust
#[open]
#[logic]
fn my_logic_function() {}

#[open]
#[logic]
pub fn my_pub_logic_function() {
	my_logic_function()
}
```
would return the following error:
```
Cannot make `"my_logic_function"` transparent in `"my_pub_logic_function"` as it would call a less-visible item.
```
because the inner function is not `pub`.

This causes issues when we have internal helper functions which we don't want to be `pub`, but which we want to be `open` to aid the SMT solvers.

I therefore believe this restriction should be lifted (similar to how you in regular Rust are allowed to `pub` a function which internally calls a private function).

----

The reason this came up is that I was updating CreuSAT to current Creusot. It utilises a scheme of having private `_inner` predicates and logic functions which take `Seq`s, `Int`s, etc, and `pub` wrappers around those functions which take regular Rust types. Previously these `_inner`s where `pub`, which lead to me using a mix of wrappers and wrappees which made refactoring harder, and tied implementation and proof implementation together more than desired.

**Comment by jhjourdan** [comment]
I'm sorry, but I feel like this is fundamentally wrong. A private function is one the clients should not be able to see private logic functions. One argument is that the client should be able to write assertions to help discharging VCs coming from the library, and these VCs contain the body of transparent definitions.

One defect with the previous design is that transparent definitions may also contain proofs, and these proofs may require private lemmas. Thanks to #1629, this should now no longer be a problem.

**Comment by sarsko** [comment]
What's your reasoning?

> A private function is one the clients should not be able to see private logic functions.

Why? Why can I not choose to make it `open` if I want? If want the client to not see it then I'll just make it less than `open`.

Again, this is how I use these things in CreuSAT:

I have private `_inner` predicates and logic functions which take `Seqs`, `Ints`, etc, and `pub` wrappers around those functions which take regular Rust types. Previously these `_inners` where pub, which lead to me using a mix of wrappers and wrappees which made refactoring harder, and tied implementation and proof implementation together more than desired.

In other words: I want the correct `pub`-ness to ensure I use the correct functions, and I want the correct `open`-ness to make the SMT solvers happy. Linking these notions makes me make everything `pub` in order to make the SMT solvers happy.

**Comment by Lysxia** [comment]
Our reasoning is that `open` makes the body of a function part of the public API of your library, that includes all of its callees, so the callees should be `pub`.

Can you explain in more detail why you want `_inner` to be private?

**Comment by jhjourdan** [comment]
My problem with your proposal is that using a private function in an open function is a way of exposing it. I.e., you tried to make some notion private to the proof and closed definitions of a module, and you exposed it. It's not desirable, because the fact that it is private means for example that you are free to change its definition without fearing to break other proofs somewhere else.

In your case, I don't really see the problem of making public your `__inner` functions that take `Seq`/`Int` values. They are really part of the interfaces of the module, so they should be public. It's quite common in a public API to have both well-behaved wrappers and not-so-well-behaved raw functions. Think of, e.g., `get_unchecked` and `index` which is a better-behaved wrapper which is preferred in client code.

Also, I don't understand why you don't define the non-`__inner` version directly without using wrappers.

If I'm not convincing, you could perhaps give a precise example?
